app/src/selfRelease/java/org/radarcns/detail/DownloadProgress.kt

```diff
@@ -0,0 +1,130 @@
+package org.radarcns.detail
+
+import android.annotation.SuppressLint
+import android.content.Context
+import okhttp3.*
+import okhttp3.Interceptor
+import okio.*
+import org.slf4j.LoggerFactory
+import java.io.File
+import java.io.IOException
+import java.lang.Exception
+
+class DownloadProgress(context: Context, delegate: TaskDelegate) {
+    private val mContext: Context = context
+    private val mDelegate: TaskDelegate = delegate
+
+    @SuppressLint("SetWorldReadable")
+    @Throws(Exception::class)
+    fun run(url: String) {
+        val request: Request = Request.Builder()
+            .url(url)
+            .build()
+        val progressListener: ProgressListener = object : ProgressListener {
+            var firstUpdate = true
+            override fun update(bytesRead: Long, contentLength: Long, done: Boolean) {
+                if (done) {
+                    logger.info("APK Download Completed")
+                } else {
+                    if (firstUpdate) {
+                        firstUpdate = false
+                        if (contentLength == -1L) {
+                            logger.info("APK content-length: unknown")
+                        } else {
+                            logger.info("APK content-length: {}", contentLength)
+                        }
+                    }
+                    if (contentLength != -1L) {
+                        mDelegate.taskProgressResult(100 * bytesRead / contentLength)
+                    }
+                }
+            }
+        }
+        val client: OkHttpClient = OkHttpClient.Builder()
+            .addNetworkInterceptor(Interceptor { chain: Interceptor.Chain ->
+                val originalResponse = chain.proceed(chain.request())
+                originalResponse.newBuilder()
+                    .body(ProgressResponseBody(originalResponse.body, progressListener))
+                    .build()
+            })
+            .build()
+        client.newCall(request).execute().use { response ->
+            if (!response.isSuccessful) throw IOException("Unexpected code $response")
+            val apkData = response.body?.byteStream()
```

Comment:
The code after this shouldn't run if there's no response body, so you can return from `use` if it is.
```suggestion
            val apkData = response.body?.byteStream() ?: run {
                logger.error("No APK data provided from ${request.url}")
                return@use
            }
```
You can remove the `if (apkData != null)` check afterwards.